_rules/invalid-form-field-value-36b590.md

```diff
@@ -0,0 +1,203 @@
+---
+id: 36b590
+name: Error message describes invalid form field value
+rule_type: atomic
+description: |
+  This rule checks that text descriptions are provided when the user completes a form field with information that is not an allowed value or using a not allowed format.
+accessibility_requirements:
+  wcag20:3.3.1: # Error Identification (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+  wcag-technique:G84: # Providing a text description when the user provides information that is not in the list of allowed values
+    forConformance: false
+    failed: not satisfied
+    passed: satisfied
+    inapplicable: further testing needed
+  wcag-technique:G85: # Providing a text description when user input falls outside the required format or values
+    forConformance: false
+    failed: not satisfied
+    passed: satisfied
+    inapplicable: further testing needed
+input_aspects:
+  - DOM Tree
+  - CSS Styling
+  - Language
+acknowledgements:
+  authors:
+    - Carlos Duarte
+    - João Vicente
+---
+
+## Applicability
+
+The rule applies to each HTML [input element][] with the [type attribute][] in one of the following [states][]: "Text", "Telephone", "URL", "E-mail", "Password", "Date", "Month", "Week", "Time", "Local Date and Time", and "Number";
```

Comment:
I don't understand why this applicability is what it is. On the one hand, it looks like you've picked things that you know have a required format. E-mail, URL, Date, etc. Someone could programatically determine that those things exist on a page, and that submitting those with an incorrect value would result in an error. But then I don't understand why text and password are on this list. Those don't have a required format, at least not always.

On the other hand, if the intent wasn't to just limit this rule to things that have a required format, then this rule might as well include all things that inherit from ARIA input or ARIA select. That would also make this required field rule redundant.

It's also worth noting that browser won't let you set invalid values for some of these fields. You can't type letters into a type=number, type=date, type=week etc. field. Browsers enforce these things, so I'm not sure what the benefit is of testing that an error is shown for them, when it's straight up not possible to fill them out incorrectly.

If you're looking for heuristics that tell you programatically whether or not a field will be validated, but that the browser doesn't handle, the only things I know of that are like that is type=email, and type=url. Autocomplete could also be a heuristic that will tell you if there's likely a format. But those are heuristics. We'd have to make an assumption that a type=email field is actually validated. That's not an assumption I would personally be comfortable with making with our tool.

Reply:
My goal was to identify form fields where a user can enter an invalid value or use an invalid format. That's why I picked the types that can be rendered by a browser as text fields (I did add a note explaining this was the criteria). 

As you said, for some of these types, browsers will prevent the user to enter invalid formats. But not all browsers will do that (Safari doesn't support type="date" for instance and falls back to type="text"). Text and password are in there for those situations where the valid format or the range of valid values is presented in the document. For example:

- Enter only your first and last names (a script can check if the user enters exactly two words)
- The password must include a lowercase letter, an uppercase letter and a number (a script will check the format)

Going with ARIA input or ARIA select would make this applicable to a lot of elements where the user cannot enter invalid values or control the format used, so in fact, they are not applicable.

Does this make sense?

Reply:
You convinced me. Changing to elements with roles that inherit from ARIA input and ARIA select.